### PR TITLE
ADBDEV-6148: Fix temporary schema cleanup after segment failure

### PR DESCRIPTION
**Pull request by silent-observer**
Fix temporary schema cleanup after segment failure

After some segment failures (such as during long insert operations) temporary
schema wasn't deleted even after the session exits. This error was caused by a
complicated interaction between two places that delete temporary tables:
RemoveTempRelationsCallback and GpDropTempTables.

This patch refactors these two functions to make their actions more clear.
GpDropTempTables is split into two functions: GpResetSessionIfNeeded and
GpDropTempTables, and RemoveTempRelationsCallaback reuses GpDropTempTables.
There is also a new function GpHasTempRelationsForDeletion that is used to
check if any cleanup is necessary. Cancelling the RemoveTempRelationsCallback
is also moved from ResetTempNamespace (which doesn't fully delete it yet, so
the callback might still be needed) to GpDropTempTables. With this the issue
is now fixed.

### CODE REVIEW

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 1]
```diff
@@ -861,18 +874,26 @@ GpDropTempTables(void)
...
+static void
+GpScheduleSessionResetInternal(bool primaryGangLoss)
 {
 	/*
- 	 * resetSessionForPrimaryGangLoss could be called twice in a transacion,
- 	 * we need to use NeedResetSession to double check if we should do the
- 	 * real work to avoid that OldTempToastNamespace be makred invalid before
- 	 * cleaning up the temp namespace.
- 	 */
+	 * GpScheduleSessionReset could be called twice in a transacion, we need
```
Comment (RekGRpth) [inline src/backend/cdb/dispatcher/cdbgang.c]: change GpScheduleSessionReset to GpScheduleSessionResetInternal in this comment and other lines of this comment may not be changed

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 2]
```diff
@@ -900,10 +921,11 @@ resetSessionForPrimaryGangLoss(void)
 			 */
 			OldTempNamespace = ResetTempNamespace();
 
-			elog(WARNING,
-				 "Any temporary tables for this session have been dropped "
-				 "because the gang was disconnected (session id = %d)",
-				 gp_session_id);
+			if (primaryGangLoss)
```
Comment (RekGRpth) [inline src/backend/cdb/dispatcher/cdbgang.c]: reduce diff by
```suggestion
			if (!primaryGangLoss)
				return;
```

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 3]
```diff
@@ -913,6 +935,32 @@ resetSessionForPrimaryGangLoss(void)
...
+{
+	GpScheduleSessionResetInternal(true);
+}
+
+/*
+ * Schedule this session for reset and register temporary tables for deletion.
+ * This function doesn't actually reset or delete anything by itself,
+ * the session will be reset on the next GpResetSessionIfNeeded call,
+ * and the temporary tables will be dropped by GpDropTempTables.
+ * Does not output a warning about dropping temporary tables.
```
Comment (RekGRpth) [inline src/backend/cdb/dispatcher/cdbgang.c:955]: same three identical comments, do something with it

File: src/backend/catalog/namespace.c [hunk 4]
```diff
@@ -4554,31 +4548,16 @@ RemoveTempRelationsCallback(int code, Datum arg)
 		return;
 	}
 
+	/* If there is a temporary namespace, schedule its deletion */
 	if (OidIsValid(myTempNamespace))
+		GpScheduleSessionReset();
```
Comment (RekGRpth) [inline src/backend/catalog/namespace.c:4558]: How are `Temp Namespace` and `Session Reset` related?

File: src/backend/catalog/namespace.c [hunk 5]
```diff
@@ -4554,31 +4548,16 @@ RemoveTempRelationsCallback(int code, Datum arg)
...
+
+	if (GpHasTempNamespaceForDeletion())
```
Comment (RekGRpth) [inline src/backend/catalog/namespace.c:4560]: may be add some comment before?

File: src/backend/cdb/dispatcher/cdbgang.c (hunk 3 above)
Comment (silent-observer) [reply src/backend/cdb/dispatcher/cdbgang.c:955]: How should that be treated in general? The `Internal` function is the one with the actual code, so it needs the comment, but the `resetSessionForPrimaryGangLoss` and `GpScheduleSessionReset` are the outwards-facing functions, so they are the ones the reader will find first. Which comments should be shortened then?

File: src/backend/catalog/namespace.c (hunk 4 above)
Comment (silent-observer) [reply src/backend/catalog/namespace.c:4558]: `GpScheduleSessionReset` also registers the temporary namespace for deletion (moves it from `myTempNamespace` to `OldTempNamespace` global variable). `TempNamespace` functions then interact with `OldTempNamespace`.

File: src/backend/catalog/namespace.c (hunk 5 above)
Comment (silent-observer) [reply src/backend/catalog/namespace.c:4560]: Something like "If we have a temporary namespace, schedule it for deletion"?

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 6]
```diff
@@ -897,9 +913,13 @@ resetSessionForPrimaryGangLoss(void)
 			 * inaccessible.  Later, when we can start a new transaction, we
 			 * will attempt to actually drop the old session tables to release
 			 * the disk space.
+			 * This will happen on the next GpDropTempTables call that isn't
+			 * inside a transaction.
 			 */
 			OldTempNamespace = ResetTempNamespace();
 
+			if (!primaryGangLoss)
+				return;
```
Comment (RekGRpth) [inline src/backend/cdb/dispatcher/cdbgang.c:922]:
```suggestion
				return;

```

File: src/backend/catalog/namespace.c (hunk 5 above)
Comment (RekGRpth) [reply src/backend/catalog/namespace.c:4560]: may be "If we have a temporary namespace, delete it"

Comment (silent-observer) [reply src/backend/catalog/namespace.c:4560]: For the `if (GpHasTempNamespaceForDeletion())` line? Okay

**Comment by RekGRpth** [comment]
reformat PR description according to guidelines

**Comment by RekGRpth** [comment]
> and `RemoveTempRelationsCallaback` reuses `GpDropTempTables`

may be "and `RemoveTempRelationsCallaback` reuses `GpDropTables`"?

or

> GpDropTempTables is split into two functions: GpResetSessionIfNeeded and GpDropTables

may be "GpDropTempTables is split into two functions: GpResetSessionIfNeeded and GpDropTempTables"?

**Comment by silent-observer** [comment]
Fixed, there is no `GpDropTables` function, that was a typo

**Review by RekGRpth — APPROVED** [review]

**Review by red1452 — COMMENTED** [review]
Look good, but there are failed tests

**Review by red1452 — APPROVED** [review]

**Pull request merged**